stl/inc/xutility

```diff
@@ -3246,10 +3259,26 @@ namespace ranges {
         }
     };
 
-    // VARIABLE ranges::distance
     inline constexpr _Distance_fn distance{_Not_quite_object::_Construct_tag{}};
 
-    // CLASS ranges::_Next_fn
+    // VARIABLE ranges::ssize
+    class _Ssize_fn { // Per the proposed resolution of LWG-3403
+    public:
+        // clang-format off
+        template <class _Rng>
+        _NODISCARD constexpr auto operator()(_Rng&& _Range) const requires requires { _RANGES size(_Range); } {
+            using _Sty = _Make_signed_like_t<decltype(_RANGES size(_Range))>;
+            using _Ty  = common_type_t<conditional_t<is_integral_v<_Sty>, ptrdiff_t, _Sty>, _Sty>;

```

Comment:
I think I'm following what's happening here, but am having some trouble tracing it back to what it says in the Standard for `ranges::ssize`.   What I found in the Standard references the width of `range_difference_t` to determine the returned type (either `range_difference_t` or `ptrdiff_t`).  Could someone explain how the `is_integral_v` comes into play here?

Reply:
The wording in the Standard is broken; WG21 (really just Jonathan Wakely and I) has been rewriting it in https://cplusplus.github.io/LWG/issue3403. Yesterday we finally landed on:

> Given a subexpression `E` with type `T`, let `t` be an lvalue that denotes the reified object for `E`. If `ranges::size(t)` is ill-formed, `ranges::ssize(E)` is ill-formed. Otherwise let `D` be `make-signed-like-t<decltype(ranges::size(t))>`, or `ptrdiff_t` if it is wider than that type; `ranges::ssize(E)` is expression-equivalent to `static_cast<D>(ranges::size(t))`.

in the reflector discussion. (I've submitted https://github.com/cplusplus/LWG/pull/332 to update the issue wording, but it hasn't merged yet.)

Reply:
As for the `is_integral_v`, my understanding of what the code is doing is:

1. Ask for what `ranges::size` would return. This is typically an unsigned integral type (typically `size_t`) but eventually it may be an implementation-defined class type imitating a large unsigned integer.
2. Use `_Make_signed_like_t` to transform that to its signed counterpart, which will be a signed integral type, or possibly a class type imitating a large signed integer.
3. Now, `ranges::ssize` would like to widen any small signed integral types to `ptrdiff_t` (this is the "or `ptrdiff_t` if it is wider than that type" from the fixed wording; I believe the assumption is that any class types will not be imitating small integers). Ideally, we could just say `common_type_t<ptrdiff_t, _Sty>` to get the wider of `ptrdiff_t` and `_Sty`. (`common_type_t` has some ridiculously complicated wording now, but it's essentially what the conditional operator `condition ? expr1 : expr2` does to convert two expressions to a single type.) However, my understanding is that the class types imitating integers may not work with `common_type_t`.
4. So, `is_integral_v` performs a bit of metaprogramming. If `_Sty` is integral, then we form `common_type_t<ptrdiff_t, _Sty>`, which (for example) will widen `int` to `ptrdiff_t`. If `_Sty` is a class type, even if it tries to imitate an integer, `is_integral_v` will be `false`, so we'll form `common_type_t<_Sty, _Sty>`. That's defined as just being `_Sty` (with some decay involved).